File: services/analytics-api/src/services/kafka-consumer.ts
import { Kafka, Consumer, KafkaMessage } from 'kafkajs';
import { EventEmitter } from 'events';
import { Order } from '../models/order';

/**
 * Kafka consumer service with intentional type inconsistency issues
 * This demonstrates the problems that occur when services use different types
 */
export class KafkaConsumerService {
  private consumer: Consumer;
  private eventEmitter: EventEmitter;
  private running: boolean = false;
  private errorCount: number = 0;
  private lastErrors: string[] = [];
  private maxErrorsToTrack: number = 10;
  
  constructor(
    private readonly brokers: string[],
    private readonly topic: string,
    private readonly groupId: string
  ) {
    const kafka = new Kafka({
      clientId: 'analytics-api',
      brokers: this.brokers
    });
    
    this.consumer = kafka.consumer({ groupId: this.groupId });
    this.eventEmitter = new EventEmitter();
  }
  
  /**
   * Start consuming messages from Kafka
   */
  async start(): Promise<void> {
    if (this.running) {
      console.log('Kafka consumer already running');
      return;
    }
    
    try {
      await this.consumer.connect();
      await this.consumer.subscribe({ topic: this.topic, fromBeginning: true });
      
      await this.consumer.run({
        eachMessage: async ({ topic, partition, message }) => {
          try {
            this.processMessage(message);
          } catch (error: any) {
            this.trackError(`Error processing message: ${error.message || 'Unknown error'}`);
            console.error('Error processing message:', error);
          }
        }
      });
      
      this.running = true;
      console.log(`Started Kafka consumer for topic ${this.topic}`);
    } catch (error) {
      console.error('Failed to start Kafka consumer:', error);
      throw error;
    }
  }
  
  /**
   * Stop consuming messages from Kafka
   */
  async stop(): Promise<void> {
    if (!this.running) {
      return;
    }
    
    try {
      await this.consumer.disconnect();
      this.running = false;
      console.log('Stopped Kafka consumer');
    } catch (error) {
      console.error('Error stopping Kafka consumer:', error);
      throw error;
    }
  }
  
  /**
   * Process a message from Kafka
   * This has intentional type conversion issues for demo purposes
   */
  private processMessage(message: KafkaMessage): void {
    if (!message.value) {
      this.trackError('Received empty message');
      return;
    }

    // Check if this is an Avro message (magic byte 0x00)
    if (message.value[0] === 0) {
      console.debug('Received Avro message, skipping in JSON consumer');
      return;
    }

    try {
      // First try to parse as JSON
      const messageStr = message.value.toString();
      const data = JSON.parse(messageStr);
      
      // Preserve types to match source data and Avro schema
      const order: Order = {
        orderId: data.orderId || '',
        // Preserve userId as string to match Avro schema and source data type
        userId: data.userId || '',
        // Type inconsistency: Convert BigDecimal to string
        amount: data.amount ? data.amount.toString() : '0',
        status: data.status || 'UNKNOWN',
        items: data.items || [],
        createdAt: data.createdAt || new Date().toISOString()
      };
      
      // Process items with type inconsistencies
      if (order.items && Array.isArray(order.items)) {
        order.items = order.items.map(item => ({
          // Type inconsistency: Convert string to number
          productId: parseInt(item.productId || '0'),
          quantity: item.quantity || 0,
          // Type inconsistency: Convert BigDecimal to number
          price: parseFloat(item.price || '0')
        }));
      }
      
      // Emit the processed order
      this.eventEmitter.emit('order', order);
      console.log(`Processed order: ${order.orderId}`);
      
    } catch (error: any) {
      // If JSON parsing fails, it might be Java serialized
      if (error instanceof SyntaxError) {
        this.trackError(`Failed to parse message as JSON: ${error.message}`);
        this.trackError('Received a message that appears to be Java serialized - cannot deserialize');
        console.error('Received a message that appears to be Java serialized - cannot deserialize');
      } else {
        this.trackError(`Error processing message: ${error.message}`);
        console.error('Error processing message:', error);
      }
      
      throw error;
    }
  }
  
  /**
   * Register a listener for order events
   */
  onOrder(listener: (order: Order) => void): void {
    this.eventEmitter.on('order', listener);
  }
  
  /**
   * Track an error message
   */
  private trackError(message: string): void {
    this.errorCount++;
    
    if (this.lastErrors.length >= this.maxErrorsToTrack) {
      this.lastErrors.shift();
    }
    
    this.lastErrors.push(message);
  }
  
  /**
   * Get the number of errors encountered
   */
  getErrorCount(): number {
    return this.errorCount;
  }
  
  /**
   * Get the last error messages
   */
  getLastErrors(): string[] {
    return [...this.lastErrors];
  }
}

File: services/inventory-service/inventory_service/consumer/kafka_consumer.py
"""
Kafka consumer for inventory service with intentional JSON deserialization issues
"""

import json
import logging
import threading
from typing import Dict, Any, Optional

from confluent_kafka import Consumer, KafkaError, KafkaException

logger = logging.getLogger(__name__)

class OrderKafkaConsumer:
    """
    Kafka consumer for order events with intentional JSON deserialization issues.
    This consumer will attempt to deserialize messages with field name mismatches.
    """
    
    def __init__(
        self, 
        bootstrap_servers: str, 
        topic: str, 
        group_id: str,
        inventory_store: Dict[str, Dict]
    ):
        self.bootstrap_servers = bootstrap_servers
        self.topic = topic
        self.group_id = group_id
        self.inventory_store = inventory_store
        self.consumer = None
        self.running = False
        self.consumer_thread = None
        self.error_count = 0
        self.last_errors = []
        self.max_errors_to_track = 10
        
    def start(self):
        """Start the Kafka consumer in a separate thread"""
        if self.running:
            logger.warning("Kafka consumer already running")
            return
            
        self.running = True
        self.consumer_thread = threading.Thread(target=self._consume_loop)
        self.consumer_thread.daemon = True
        self.consumer_thread.start()
        logger.info(f"Started Kafka consumer for topic {self.topic}")
        
    def stop(self):
        """Stop the Kafka consumer"""
        self.running = False
        if self.consumer_thread:
            self.consumer_thread.join(timeout=5.0)
            self.consumer_thread = None
        if self.consumer:
            self.consumer.close()
            self.consumer = None
        logger.info("Stopped Kafka consumer")
        
    def _create_consumer(self):
        """Create a new Kafka consumer instance"""
        config = {
            'bootstrap.servers': self.bootstrap_servers,
            'group.id': self.group_id,
            'auto.offset.reset': 'earliest',
            'enable.auto.commit': True,
        }
        return Consumer(config)
        
    def _consume_loop(self):
        """Main consumer loop"""
        try:
            self.consumer = self._create_consumer()
            self.consumer.subscribe([self.topic])
            
            while self.running:
                msg = self.consumer.poll(timeout=1.0)
                
                if msg is None:
                    continue
                    
                if msg.error():
                    if msg.error().code() == KafkaError._PARTITION_EOF:
                        logger.debug(f"Reached end of partition {msg.partition()}")
                    else:
                        logger.error(f"Kafka error: {msg.error()}")
                    continue
                    
                try:
                    self._process_message(msg.value())
                except Exception as e:
                    self.error_count += 1
                    error_msg = str(e)
                    self.last_errors.append(error_msg)
                    # Keep only the last N errors
                    if len(self.last_errors) > self.max_errors_to_track:
                        self.last_errors.pop(0)
                    logger.error(f"Error processing message: {e}")
                    
        except KafkaException as e:
            logger.error(f"Kafka exception: {e}")
        finally:
            if self.consumer:
                self.consumer.close()
                
    def _process_message(self, message_bytes: bytes):
        """
        Process a message from Kafka.
        Intentionally uses different field names than the producer to demonstrate deserialization issues.
        """
        try:
            # Check if this is an Avro message (magic byte 0x00)
            if len(message_bytes) > 0 and message_bytes[0] == 0:
                logger.debug("Received Avro message, skipping in JSON consumer")
                return

            # First try to deserialize as JSON
            try:
                # Attempt to decode as JSON
                message_str = message_bytes.decode('utf-8')
                order_data = json.loads(message_str)
                
                # Intentional field name mismatches for demo purposes
                # The producer uses 'orderId', but we're looking for 'order_id'
                # The producer uses 'userId', but we're looking for 'user_id'
                order_id = self._extract_field(order_data, ['order_id', 'orderId', 'orderid'])
                user_id = self._extract_field(order_data, ['user_id', 'userId', 'userid'])
                
                if not order_id:
                    raise ValueError("Could not find order_id field in message")
                
                # Process items with field name mismatches
                items = self._extract_field(order_data, ['items', 'orderItems', 'order_items'], [])
                
                # Update inventory store
                for item in items:
                    product_id = self._extract_field(item, ['product_id', 'productId', 'productid'])
                    quantity = self._extract_field(item, ['quantity', 'qty', 'amount'])
                    
                    if not product_id or quantity is None:
                        logger.warning(f"Skipping item with missing fields: {item}")
                        continue
                        
                    inventory_entry = {
                        "order_id": order_id,
                        "product_id": product_id,
                        "quantity": quantity,
                        "status": "RESERVED",
                        "user_id": user_id
                    }
                    
                    # Store by order_id for retrieval
                    self.inventory_store[order_id] = inventory_entry
                    logger.info(f"Updated inventory for order {order_id}, product {product_id}")
                
            except (UnicodeDecodeError, json.JSONDecodeError) as e:
                # If JSON deserialization fails, try to handle as Java serialized object
                # This will fail, but we want to show the error for demo purposes
                logger.error(f"Failed to deserialize message as JSON: {e}")
                logger.error("Received a message that appears to be Java serialized - cannot deserialize")
                raise ValueError("Received a message that appears to be Java serialized - cannot deserialize")
                
        except Exception as e:
            logger.error(f"Error processing message: {e}")
            raise
            
    def _extract_field(self, data: Dict[str, Any], possible_names: list, default=None) -> Optional[Any]:
        """
        Try to extract a field from the data using multiple possible field names.
        This is to demonstrate field name mismatch issues.
        """
        for name in possible_names:
            if name in data:
                return data[name]
        return default
        
    def get_error_count(self) -> int:
        """Get the number of errors encountered"""
        return self.error_count
    
    def get_last_errors(self) -> list:
        """Get the list of last error messages"""
        return self.last_errors.copy()
